File: algorithms/1_easy/P/01013_partition_array_into_three_parts_with_equal_sum/solution.c
bool canThreePartsEqualSum(const int *restrict arr, const int len) {
    if (len < 3 || arr == NULL) {
        return false;
    }

    int sum = 0;
    for (int i = 0; i < len; i += 1) {
        sum += arr[i];
    }
    if ((sum % 3) != 0) {
        return false;
    }

    const int target = sum / 3;
    int cnt = 0;
    int curr_sum = 0;
    for (int i = 0; i < len; i += 1) {
        curr_sum += arr[i];
        if (curr_sum == target) {
            cnt += 1;
            if (cnt == 3) {
                return true;
            }
            curr_sum = 0;
        }
    }

    return false;
}

File: algorithms/1_easy/P/01013_partition_array_into_three_parts_with_equal_sum/solution.rs
impl Solution {
    pub fn can_three_parts_equal_sum(arr: Vec<i32>) -> bool {
        let sum: i32 = arr.iter().sum();
        if (sum % 3) != 0 {
            return false;
        }

        let target = sum / 3;
        let mut cnt = 0;
        let mut curr_sum = 0;
        for a in arr {
            curr_sum += a;
            if curr_sum == target {
                cnt += 1;
                if cnt == 3 {
                    return true;
                }
                curr_sum = 0;
            }
        }

        false
    }
}
